triggers/service/auth/src/main/scala/com/daml/oauth/server/Request.scala

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.oauth.server
+
+import akka.http.scaladsl.model.Uri
+import akka.http.scaladsl.model.Uri.Query
+import spray.json.{
+  DefaultJsonProtocol,
+  JsString,
+  JsValue,
+  JsonFormat,
+  RootJsonFormat,
+  deserializationError
+}
+
+object Request {
+
+  // https://tools.ietf.org/html/rfc6749#section-4.1.1
+  case class Authorize(
+      responseType: String,
+      clientId: String,
+      redirectUri: Uri, // optional in oauth but we require it
+      scope: Option[String],
+      state: Option[String]) {
+    def toQuery: Query = {
+      var params: Seq[(String, String)] =
+        Seq(
+          ("response_type", responseType),
+          ("client_id", clientId),
+          ("redirect_uri", redirectUri.toString))
+      scope.foreach { scope =>
+        params ++= Seq(("scope", scope))
+      }
+      state.foreach { state =>
+        params ++= Seq(("state", state))
+      }
+      Query(params: _*)
+    }
+  }
+
+  // https://tools.ietf.org/html/rfc6749#section-4.1.3
+  case class Token(
+      grantType: String,
+      code: String,
+      redirectUri: Option[Uri],
```

Comment:
https://tools.ietf.org/html/rfc6749#section-4.1.3 says it's required if present in the request, why is it optional here?

Reply:
Good point, that was when I still had it optional in the request as well which turned out to be annoying.